libraries/AP_Proximity/AP_Proximity_LeddarVu8.cpp

```diff
@@ -0,0 +1,247 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_Proximity_LeddarVu8.h"
+#include <AP_HAL/AP_HAL.h>
+#include <ctype.h>
+#include <GCS_MAVLink/GCS.h>
+#include <GCS_MAVLink/GCS_MAVLink.h>
+#if HAL_PROXIMITY_ENABLED && AP_PROXIMITY_LEDDARVU8_ENABLED
+
+// update the state of the sensor
+void AP_Proximity_LeddarVu8::update()
+{
+    if (!_initialized) {
+        request_distances();
+        // gcs().send_text(MAV_SEVERITY_WARNING,"requested distances");
+        _temp_boundary.reset();
+        _initialized = true;
+        last_distance_ms = AP_HAL::millis();
+    }
+
+    if ((AP_HAL::millis() - last_distance_ms) < LEDDARVU8_TIMEOUT_MS) {
+        // just initialized
+        set_status(AP_Proximity::Status::Good);
+    } else {
+        set_status(AP_Proximity::Status::NoData);
+        // gcs().send_text(MAV_SEVERITY_WARNING,"no data");
+    }
+
+    // read data
+    read_sensor_data();
+    // gcs().send_text(MAV_SEVERITY_WARNING,"reading data");
+
+    if (AP_HAL::millis() - last_distance_ms < LEDDARVU8_TIMEOUT_MS) {
+        set_status(AP_Proximity::Status::Good);
+        // gcs().send_text(MAV_SEVERITY_WARNING,"good data");
+    } else {
+        // long time since we received any valid sensor data
+        // try sending the sensor the "send data" message
+        _initialized = false;
+        set_status(AP_Proximity::Status::NoData);
+        // gcs().send_text(MAV_SEVERITY_WARNING,"no data in the else");
+    }
+}
+// extern const AP_HAL::HAL& hal;
+
+// LeddarVu8 uses the modbus RTU protocol
+// https://autonomoustuff.com/product/leddartech-vu8/
+// distance returned in reading_cm, signal_ok is set to true if sensor reports a strong signal
+void AP_Proximity_LeddarVu8::read_sensor_data()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+
+    // check for timeout receiving messages
+    uint32_t now_ms = AP_HAL::millis();
+    if (((now_ms - last_distance_ms) > LEDDARVU8_TIMEOUT_MS) && ((now_ms - last_distance_request_ms) > LEDDARVU8_TIMEOUT_MS)) {
+        request_distances();
+    }
+
+    // read any available characters from the lidar
+    int16_t nbytes = _uart->available();
+    while (nbytes-- > 0) {
+        int16_t r = _uart->read();
```

Comment:
Maybe something like this is little cleaner?
```suggestion
        const uint8_t byte = _uart->read();
```